miio/airhumidifier_jsq.py

```diff
@@ -0,0 +1,330 @@
+        token: str = None,
+        start_id: int = 0,
+        debug: int = 0,
+        lazy_discover: bool = True,
+        model: str = MODEL_HUMIDIFIER_JSQ001,
+    ) -> None:
+        super().__init__(ip, token, start_id, debug, lazy_discover)
+
+        if model in AVAILABLE_PROPERTIES:
+            self.model = model
+        else:
+            self.model = MODEL_HUMIDIFIER_JSQ001
+
+        self.device_info = None
+
+    @command(
+        default_output=format_output(
+            "",
+            "Power: {result.power}\n"
+            "Mode: {result.mode}\n"
+            "Temperature: {result.temperature} °C\n"
+            "Humidity: {result.humidity} %\n"
+            "Buzzer: {result.buzzer}\n"
+            "LED brightness: {result.led_brightness}\n"
+            "Child lock: {result.child_lock}\n"
+            "No water: {result.no_water}\n"
+            "Lid opened: {result.lid_opened}\n"
+            "Firmware version: {result.firmware_version}\n"
+            "Hardware version: {result.hardware_version}\n",
+        )
+    )
+    def status(self) -> AirHumidifierStatus:
+        """Retrieve properties."""
+
+        if self.device_info is None:
+            self.device_info = self.info()
+
+        values = self.send("get_props")
+
+        # Response of an Air Humidifier (shuii.humidifier.jsq001):
+        # [24, 37, 3, 1, 0, 2, 0, 0, 0]
+        #
+        # status[0] : temperature (degrees, int)
+        # status[1]: humidity (percentage, int)
+        # status[2]: mode ( 0: Intelligent, 1: Level1, ..., 5:Level4)
+        # status[3]: buzzer (0: off, 1: on)
+        # status[4]: lock (0: off, 1: on)
+        # status[5]: brightness (0: off, 1: low, 2: high)
+        # status[6]: power (0: off, 1: on)
+        # status[7]: water level state (0: ok, 1: add water)
+        # status[8]: lid state (0: ok, 1: lid is opened)
+
+        properties = AVAILABLE_PROPERTIES[self.model]
+        if len(properties) != len(values):
+            _LOGGER.error(
+                "Count (%s) of requested properties (%s) does not match the "
+                "count (%s) of received values (%s).",
+                len(properties),
+                properties,
+                len(values),
+                values,
+            )
+
+        return AirHumidifierStatus(
+            {k: v for k, v in zip(properties, values)}, self.device_info
+        )
+
+    @command(default_output=format_output("Powering on"))
+    def on(self):
+        """Power on."""
+        return self.send("set_start", [1])
+
+    @command(default_output=format_output("Powering off"))
+    def off(self):
+        """Power off."""
+        return self.send("set_start", [0])
+
+    @command(
+        click.argument("mode", type=EnumType(OperationMode, False)),
+        default_output=format_output("Setting mode to '{mode.value}'"),
+    )
+    def set_mode(self, mode: OperationMode):
+        """Set mode."""
+        value = mode.value
+        if value not in (
```

Comment:
I'm wondering if this could be simplified by simply testing against all the name in the enum? Something based on `list(OperationModel)`?

Reply:
The most evident way like `mode in OperationMode` is deprecated in 3.8
```
>>> OperationMode
<enum 'OperationMode'>
>>> 0 in OperationMode
__main__:1: DeprecationWarning: using non-Enums in containment checks will raise TypeError in Python 3.8
False
>>> 8 in OperationMode
False
```

List doesn't work here
```
>>> list(OperationMode)
[<OperationMode.Intelligent: 0>, <OperationMode.Level1: 1>, <OperationMode.Level2: 2>, <OperationMode.Level3: 3>, <OperationMode.Level4: 4>]
>>> 0 in list(OperationMode)
False
```

But we could use a generator. I agree it is much better that hardcoded enum options.
```
>>> [v.value for v in OperationMode]
[0, 1, 2, 3, 4]
```

Reply:
Fixed, see the updated version.